posts/virtual_environments.md

```diff
@@ -0,0 +1,126 @@
+
+![Figure 3: Environments managed through the virtual layer](virtual_environments/virtual_envs.png)
+*Figure 3: Environments managed through the virtual layer.*
+
+These two properties combined is what constitutes **virtual environments**.
+
+### Model Snapshots
+As mentioned earlier, every time a model changes a new model snapshot gets created to capture the change. Each snapshot represents the state of a model at the time when this snapshot was generated. In SQLMesh snapshots are generated automatically when a new [plan](https://sqlmesh.readthedocs.io/en/stable/concepts/plans/) is created and applied.
+
+#### Fingerprinting
+Each snapshot is uniquely identified by its "fingerprint". A fingerprint consists of the following attributes:
+- The data hash of the model is based on model attributes that have a direct impact on the dataset produced by this model. Such attributes include the model's query, storage format, partitioning scheme, etc. When this hash changes the model is considered to be modified **directly**.
+- The data hash of upstream models. This is the same data hash calculation logic but applied to models that are upstream from the current model in order to capture data changes caused by the model's dependencies. Similarly, when this hash changes the model is considered to be modified **indirectly**.
+- The metadata hash of the model is based on model attributes that have no impact on data itself. This includes metadata like ownership information, descriptions, comments, etc.
+- The metadata hash of upstream models captures metadata changes in model's dependencies.
+
+Additionally, thanks to the [SQLGlot optimizer](https://github.com/tobymao/sqlglot/blob/main/posts/python_sql_engine.md#optimizing) we’re able to create a canonical version of each SQL query, so that non-functional changes like formatting or structural refactoring don’t impact the fingerprint.
+
+#### Automatic Change Categorization
+The difference between **direct** and **indirect** modifications becomes relevant when [categorizing](https://sqlmesh.readthedocs.io/en/stable/concepts/plans/#change-categories) individual changes during plan creation.
+
+Every time a model is modified directly, SQLMesh automatically categorizes the change as “breaking” or “non-breaking” based on the change’s impact on models that are downstream from the modified one. This is possible thanks to the platform’s ability to understand SQL semantically, at the [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) level. By comparing the current model version with the previous one we’re able to compute a [semantic diff](https://github.com/tobymao/sqlglot/blob/main/posts/sql_diff.md) which SQLMesh further analyzes to understand what kind of impact it would have on downstream models.
```

Comment:
"breaking" (downstream models are affected) or "non-breaking" (only this model is affected)